### PR TITLE
MM-18920 Migrate 'utils/url.jsx' and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by DropNib**
#### Summary
This pull request migrates utils/url.jsx and utils/url.test.jsx to TypeScript.

One part where I wasn't sure was the [forEach loop](https://github.com/mattermost/mattermost-webapp/compare/master...DropNib:MM-18920?expand=1#diff-69853f2bd39d0d0041acccd6a871e0c0R45) in the test file.
I got two eslint errors that there were to many nested callbacks on this line. As I see it, the existing nested loop is elegant and easy to understand and trying to reduce the nesting will only result in less readable code. So I disabled the eslint rule for this line.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/12483

### CODE REVIEW

File: utils/url.tsx [hunk 1]
```diff
@@ -3,23 +3,33 @@
 
 import {latinise} from 'utils/latinise.jsx';
 
-export function cleanUpUrlable(input) {
-    var cleaned = latinise(input);
+type WindowObject = {
+    location: {
```
Comment (devinbinnie) [inline utils/url.tsx]: You can use the provided `Location` type here.
https://developer.mozilla.org/en-US/docs/Web/API/Location

Comment (DropNib) [reply utils/url.tsx]: Thanks @devinbinnie, I didn't know of that type.

But changing the type to that would require every caller to pass and object which fully satisfies the Location type, even though we are only interested in certain properties of it. For instance, the tests in `utils/url.test.tsx` would have to be heavily modified to implement that interface, but we wouldn't use any additional properties of it.

An alternate approach could be to make `WindowObject.location` a `Partial<Location>`, but then our logic would have to be updated to check every single property for its existence.
Of course we could use mapped types to create a Location type where all properties except the ones we are interested in are optional, but that would some hard to read code along the lines of:

```ts
type OptionalExceptFor<T, TRequired extends keyof T> = Partial<T> & Pick<T, TRequired>

type CustomLocation = OptionalExceptFor<Location, 'origin' | 'protocol' | 'hostname' | 'port'>;
```
(code adapted from [this](https://stackoverflow.com/questions/52703321/make-some-properties-optional-in-a-typescript-type) SO thread)

If constructs like that `OptionalExceptFor` are already in use in the project, then I would of course do it that way, but I haven't found one yet.

What are your thoughts on that?

File: utils/url.test.tsx [hunk 2]
```diff
@@ -42,7 +42,7 @@ describe('Utils.URL', () => {
             },
         ];
 
-        testCases.forEach((testCase) => it(testCase.description, () => {
+        testCases.forEach((testCase) => it(testCase.description, () => { // eslint-disable-line max-nested-callbacks
```
Comment (devinbinnie) [inline utils/url.test.tsx]: This can be removed with the latest master branch, the `eslintrc.json` has been updated

File: utils/url.tsx (hunk 1 above)
Comment (devinbinnie) [reply utils/url.tsx]: That seems like a lot of effort, let's just go with your defined type for now.

File: utils/url.test.tsx (hunk 2 above)
Comment (DropNib) [reply utils/url.test.tsx]: Excellent, that's very helpful - thanks for pointing out. I updated the PR.

File: utils/url.tsx (hunk 1 above)
Comment (DropNib) [reply utils/url.tsx]: Great, thanks.

**Comment by DropNib** [comment]
> Looks great! Thanks so much @DropNib!

Thank you very much as well @devinbinnie

**Pull request merged**